programs/utilities/spellbook.sc

```diff
@@ -0,0 +1,526 @@
+  book:'vbook' = book:'vbook' + 1;
+
+  write_file('books/'+_sanitize_book_title(book:'title'), 'json', book);
+);
+
+_write_cache(title, book_nbt) -> (
+  write_file('cache/'+_sanitize_book_title(title), 'nbt', book_nbt);
+);
+
+_read_cache(title) -> (
+  return(read_file('cache/'+_sanitize_book_title(title), 'nbt'));
+);
+
+
+
+// Because part of the in game book nbt is used to write a file 
+// remove any weird characters from the spellbook title.
+// This is probably a non-issue because 
+// scarpet seems to handles file security well, but just in case.
+// This would present a issue for non-roman characters. 
+_sanitize_book_title(title) -> (
+  return(replace(title, '[^A-z0-9_-]', '_'));
+);
+
+// Command Methods
+help(page) -> (
+  print(player(), format(global_help_pages:page));
+);
+
+// Player bot shorthand
+set_bot_at_player(book, spell, bot_name) -> (
+  p = player();
+  set_bot( book, spell, bot_name, map(p~'pos', round(_)), p~'dimension');
+);
+
+set_bot(book, spell, bot_name, location, dimension) -> (
+  location = join(' ', location);
+  _set_commands( book, [{
+    'title'-> spell+' +',
+    'command'-> str('/player %s spawn at %s facing 1 1 in %s', bot_name, location, dimension ),
+    'tooltip'-> str('Spawn %s bot at %s in %s', bot_name, location, dimension)
+  },{
+    'title'-> spell+' -',
+    'command'-> str('/player %s kill', bot_name),
+    'tooltip'-> str('Kill %s bot in %s', bot_name, dimension)
+  }]);
+);
+
+// forceload shorthand
+set_forceload_at_player(book, spell, from, to) -> (
+  set_forceload(book, spell, from, to, player()~'dimension');
+);
+
+set_forceload(book, spell, from, to, dimension) -> (
+  from = join(' ', from);
+  to = join(' ', to);
+  _set_commands( book, [{
+    'title'->spell+' +', 
+    'command'-> str('/execute in %s run forceload add %s %s', dimension, from, to),
+    'tooltip'-> str('forceload chunks from %s to %s in %s', from, to, dimension)
+  },{
+    'title'-> spell+' -', 
+    'command'-> str('/execute in %s run forceload remove %s %s', dimension, from, to),
+    'tooltip'-> str('unload chunks from %s to %s in %s', from, to, dimension)
+  }]);
+);
+
+// Warp Shorthand
+set_warp_at_player(book, spell) -> (
+  p = player();
+  set_warp(book, spell, map(p~'pos', round(_)), p~'dimension');
+);
+
+set_warp(book, spell, location, dimension) -> (
+  location = join(' ', location);
+  _set_commands( book, [{
+    'title'-> spell, 
+    'command'-> str('/execute as @p in %s run tp %s', dimension, location ),
+    'tooltip'-> str('warp to %s in %s', location, dimension)
+  }]);
+);
+
+
+
+list_books() -> (
+  p = player();
+  display = [];
+  
+  for( _get_book_list(),
+    display += str('m %s', _, -1);
+    display += 'g , ';
+  );
+  
+  delete(display, -1);
+
+  _print_title(p, 'Spellbook List');
+  print(p, format( display ));
+);
+
+delete_book(book) -> (
+  delete_file('books/'+book, 'json');
+  _print_message(player(), str('Burned the %s spellbook to ashes.', book));
+);
```

Comment:
Maybe add a confirmation message to the delete action. This is very deffinitive to be able to do it by mistake.

Reply:
Yeah that's a good idea